File: 0037/0037.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        rows = [[0]*10 for _ in range(9)]
        cols = [[0]*10 for _ in range(9)]
        boxs = [[0]*10 for _ in range(9)]
        
        for i in range(9):
            for j in range(9):
                if board[i][j] != ".":
                    n = int(board[i][j])
                    rows[j][n] = 1
                    cols[i][n] = 1
                    boxs[i//3*3+j//3][n] = 1
        
        def fill(board, i, j):
            if j == 9: return True
            nxt_i = (i+1) % 9
            nxt_j = j+1 if nxt_i == 0 else j
            if board[i][j] != ".": return fill(board, nxt_i, nxt_j)
            for n in range(1, 10):
                if not rows[j][n] and not cols[i][n] and not boxs[i//3*3+j//3][n]:
                    board[i][j] = str(n)
                    rows[j][n] = 1
                    cols[i][n] = 1
                    boxs[i//3*3+j//3][n] = 1
                    if fill(board, nxt_i, nxt_j): return True
                    board[i][j] = "."
                    rows[j][n] = 0
                    cols[i][n] = 0
                    boxs[i//3*3+j//3][n] = 0
            return False
        
        fill(board, 0, 0)

File: notes/0037/0037.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        vector<vector<int>> row(9, vector<int>(10));
        vector<vector<int>> col(9, vector<int>(10));
        vector<vector<int>> box(9, vector<int>(10));
        
        for (int i=0; i<9; i++) 
            for (int j=0; j<9; j++) {
                if (board[i][j] != '.') {
                    row[i][board[i][j]-'0'] = 1;
                    col[j][board[i][j]-'0'] = 1;
                    box[3*(i/3)+j/3][board[i][j]-'0'] = 1;
                }
            }
        
        dfs(board, 0, 0, row, col, box);
    }
private:
    
    bool dfs(vector<vector<char>>& board, int i, int j, vector<vector<int>>& row, vector<vector<int>>& col, vector<vector<int>>& box) {
        if (i == 9) return true;
        
        int nxt_j = (j+1) % 9;
        int nxt_i = (nxt_j == 0) ? (i + 1) : i;
        
        if (board[i][j] != '.') return dfs(board, nxt_i, nxt_j, row, col, box);
        
        for (int n=1; n<10; n++){
            if (!row[i][n] && !col[j][n] && !box[i/3*3+j/3][n]) {
                row[i][n] = 1;
                col[j][n] = 1;
                box[i/3*3+j/3][n] = 1;
                board[i][j] = '0'+n;
                if (dfs(board, nxt_i, nxt_j, row, col, box)) return true;
                board[i][j] = '.';
                row[i][n] = 0;
                col[j][n] = 0;
                box[i/3*3+j/3][n] = 0;
            }
        }
        return false;
    }
};
